### PR TITLE
Implement corpus transformations

### LINKED ISSUE #9: Corpus transforms
We should be able to support lazy (perform at read) transformation of the corpus

```python
# Word frequency after lower case
corpus.lower().freq("words")
# Word frequency based on reversed words
corpus.transform("text", lambda text[::-1]).freq("words")
```

### PR DESCRIPTION
**Pull request by jmccrae**
Fixes #9

### CODE REVIEW

File: teanga/document.py [hunk 1]
```diff
@@ -295,6 +303,11 @@ def __getitem__(self, key):
         """Return the annotation with the given index."""
         return self.raw[key]
 
+    @abstractmethod
```
Comment (AdeDoyle) [inline teanga/document.py:306]: Any Layer subclass must implement a transform method?

File: tests/test_corpus.py [hunk 2]
```diff
@@ -125,8 +125,9 @@ def test_ud():
 
 def test_open_url():
     corpus = teanga.from_url("https://teanga.io/examples/ex1.yaml")
-    assert len(corpus.docs) == 1
-    assert corpus.docs[0][1].text.text[0] == "Teanga2 data model"
+    corpus_docs = list(corpus.docs)
```
Comment (AdeDoyle) [inline tests/test_corpus.py:128]: I'm guessing this caused trouble before I got to reviewing the files because `corpus` is now a dict object.

**Review by AdeDoyle — COMMENTED** [review]
Some of the transformations create extra data for the whole corpus, which might be better to generate on the fly as required rather than storing it (in memory?). All of the code seems to work as intended. Was there anything specific you wanted me to address before approving this?

File: teanga/corpus.py [hunk 3]
```diff
@@ -434,6 +437,63 @@ def apply(self, service : Service):
         for _, doc in self.docs:
             service.execute(doc)
 
+
+    def lower(self) -> TransformedCorpus:
```
[the remark this answers is not in the document]
Comment (jmccrae) [reply teanga/corpus.py:441]: This is the intention of the implementation. The function is applied as required

File: teanga/document.py [hunk 4]
```diff
@@ -58,6 +63,9 @@ def __setitem__(self, name:str, value) -> 'Layer':
             raise Exception("Layer with name " + name + " does not exist.")
         if value is None and self._meta[name].default is not None:
             value = self._meta[name].default
+        if isinstance(value, Layer):
```
[the remark this answers is not in the document]
Comment (jmccrae) [reply teanga/document.py:66]: Docstring should probably be updated... it is quite vague to allow for lots of different kinds of values to be passed

File: teanga/document.py (hunk 1 above)
Comment (jmccrae) [reply teanga/document.py:306]: Yes, different subclasses store data differently so they must transform differently

**Pull request merged**